source/extensions/filters/http/common/aws/signer_impl.cc

```diff
@@ -0,0 +1,145 @@
+  // Phase 1: Create a canonical request
+  const auto canonical_headers = Utility::canonicalizeHeaders(headers);
+  const auto signing_headers = createSigningHeaders(canonical_headers);
+  const auto canonical_request =
+      createCanonicalRequest(message, canonical_headers, signing_headers, content_hash);
+  ENVOY_LOG(debug, "Canonical request:\n{}", canonical_request);
+  // Phase 2: Create a string to sign
+  const auto credential_scope = createCredentialScope(short_date);
+  const auto string_to_sign = createStringToSign(canonical_request, long_date, credential_scope);
+  ENVOY_LOG(debug, "String to sign:\n{}", string_to_sign);
+  // Phase 3: Create a signature
+  const auto signature =
+      createSignature(credentials.secretAccessKey().value(), short_date, string_to_sign);
+  // Phase 4: Sign request
+  const auto authorization_header = createAuthorizationHeader(
+      credentials.accessKeyId().value(), credential_scope, signing_headers, signature);
+  ENVOY_LOG(debug, "Signing request with: {}", authorization_header);
+  headers.addCopy(Http::Headers::get().Authorization, authorization_header);
+}
+
+std::string SignerImpl::createContentHash(Http::Message& message) const {
+  if (!message.body()) {
+    return SignatureConstants::get().HashedEmptyString;
+  }
+  return Hex::encode(Ssl::Utility::getSha256Digest(*message.body()));
+}
+
+std::string SignerImpl::createCanonicalRequest(
+    Http::Message& message, const std::map<std::string, std::string>& canonical_headers,
+    absl::string_view signing_headers, absl::string_view content_hash) const {
+  std::vector<absl::string_view> parts;
+  const auto* method_header = message.headers().Method();
+  if (method_header == nullptr || method_header->value().empty()) {
+    throw EnvoyException("Message is missing :method header");
+  }
+  parts.emplace_back(method_header->value().getStringView());
+  const auto* path_header = message.headers().Path();
+  if (path_header == nullptr || path_header->value().empty()) {
+    throw EnvoyException("Message is missing :path header");
+  }
+  // don't include the query part of the path
+  const auto path = StringUtil::cropRight(path_header->value().getStringView(), "?");
+  parts.emplace_back(path.empty() ? "/" : path);
+  const auto query = StringUtil::cropLeft(path_header->value().getStringView(), "?");
+  // if query == path, then there is no query
+  parts.emplace_back(query == path ? "" : query);
+  std::vector<std::string> formatted_headers;
+  formatted_headers.reserve(canonical_headers.size());
+  for (const auto& header : canonical_headers) {
+    formatted_headers.emplace_back(fmt::format("{}:{}", header.first, header.second));
+    parts.emplace_back(formatted_headers.back());
+  }
+  // need an extra blank space after the canonical headers
+  parts.emplace_back("");
+  parts.emplace_back(signing_headers);
+  parts.emplace_back(content_hash);
+  return absl::StrJoin(parts, "\n");
+}
```

Comment:
This one might also belong in the utility like canonical headers creation, it has some edge cases that would benefit from unit tests unless already well covered.

Reply:
I do have tests in the signer_tests specifically for this function but I can pull it out into a utility. I'm not sure what the convention is on util creation in envoy. Is it for any one function that is sufficiently complex on its own? Or is it just about reuse? I don't really see the request or header canonicalization being used outside of signing AWS requests.

Reply:
It's kind of up to you largely, as long as you can provide adequate test coverage. The philosophy is at a bare minimum we have code coverage based on lines, but for complex/tricky code, it's often useful to ensure we can do behavioral/branch/functional coverage via unit tests. Some folks use testing peers, others code structure to do this. I personally don't like moving private to public methods in a class just for test, but I do like adding a logically separate utility class that has the methods as public where it makes sense.